Improve rc filter updating and initialization

Quick-Flash opened the pull request:
This should speedup some of the RC filtering updating. Instead of needing to redundantly calculate the same cutoff and pt3 filter k for 3 difference filters on each axis (9 times) we can instead just calculate the pt3 filter k value once and then apply that k to each filter, saving many calculations. I have also reduced the amount of checks for if initialized. Now feedforward cutoff and setpoint cutoff are always the same (was already the same in 95+% of setups anyhow).

Also fix the issue where the feedforward_smooth_factor would be dependent on the rx rate (ie the filter cutoff would change based on the looprate).

Bonus points, I added faster ptxFilterGainFromDelay functions, so now if you ever want to calculate a filter using a tau instead of a cutoff and update it on the fly it'll be faster to run than it was previously.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Refactor**
  - RC smoothing and feedforward reorganized into per‑axis adaptive smoothing with centralized cutoff updates and explicit smoothing‑update logic.

- **Bug Fixes / Behavior**
  - RX timing relaxed for more reliable ~1 kHz handling; system now maintains and reports a smoothed RX rate (float) used across CLI/telemetry and smoothing.

- **Chores**
  - Feedforward cutoff parameter removed/merged into setpoint cutoff; default feedforward smoothing increased for gentler response.

- **UI / Warnings**
  - RC‑smoothing init beep and on‑screen warning removed; CLI/blackbox outputs updated to show smoothed RX rate.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Quick-Flash commented on src/main/fc/rc.c:
```diff
@@ -341,61 +341,83 @@ static FAST_CODE_NOINLINE void rcSmoothingSetFilterCutoffs(rcSmoothingFilter_t *
     const uint16_t minCutoffHz = 15; // don't let any RC smoothing filter cutoff go below 15Hz
-    if (smoothingData->setpointCutoffSetting == 0) {
+
+    const bool autoSetpointSmoothing = smoothingData->setpointCutoffSetting == 0;
+    const bool autoThrottleSmoothing = smoothingData->throttleCutoffSetting == 0;
+    const bool autoFeedforwardSmoothing = smoothingData->throttleCutoffSetting == 0;
+    const bool autoSetpointAndFeedforward = autoSetpointSmoothing && autoFeedforwardSmoothing;
```
Good catch.

Quick-Flash commented on src/main/fc/rc.c:
```diff
@@ -341,61 +341,83 @@
-                pt3FilterUpdateCutoff(&smoothingData->filterFeedforward[i], pt3FilterGain(feedforwardCutoffFrequency, dT));
+
+        // Update the Feedforward filter
+        if (smoothingData->feedforwardCutoffFrequency != oldFeedforwardCutoff) {
+            pt3FilterUpdateCutoff(&smoothingData->filterFeedforward[0], pt3FilterGain(setpointCutoffFrequency, dT));
+            const float pt3K = smoothingData->filterSetpoint[0].k;
+            for (int i = FD_ROLL; i <= FD_YAW; i++) {
+                smoothingData->filterFeedforward[i].k = pt3K;
             }
```
I already have a comment mentioning that all cutoffs are the same, so maintainability should be fine for the moment.

ledvinap commented on src/main/common/filter.c:
```diff
@@ -57,8 +57,7 @@ float pt1FilterGainFromDelay(float delay, float dT)
         return 1.0f; // gain = 1 means no filtering
     }
 
-    const float cutoffHz = 1.0f / (2.0f * M_PIf * delay);
```
Is this really necessary? IMO generated code would be identical.
At least leave cutoff calculation as comment

ledvinap commented on src/main/fc/rc.c:
```diff
@@ -341,71 +341,65 @@ static FAST_CODE_NOINLINE void rcSmoothingSetFilterCutoffs(rcSmoothingFilter_t *
+        pt3FilterUpdateCutoff(&smoothingData->filterSetpoint[3], pt3FilterGain(throttleCutoffFrequency, dT));
+    }
+
+    // note that cutoff frequencies are integers, filter cutoffs won't re-calculate until there is > 1hz variation from previous cutoff
```
Isn't it better to use `float` and explicit delta?  
float value can be close to inteder, so small noise will cause frequent recalculation - nondeterministic additional CPU load can have consequences ...

Quick-Flash replied:
I'm fine making it a float And giving it a 1hz range for recalculating. I'm fine even recalculating every loop and not checking.

ledvinap replied:
Personally I'd use each cycle recalc - it's simpler and more deterministic.

Quick-Flash replied:
Perfect, done and done, more code to delete that way anyhow :)

Quick-Flash replied:
@ctzsnooze check out this conversation

ledvinap commented on src/main/fc/rc.c:
```diff
@@ -341,71 +341,65 @@ static FAST_CODE_NOINLINE void rcSmoothingSetFilterCutoffs(rcSmoothingFilter_t *
+    if (smoothingData->setpointCutoffFrequency != oldSetpointCutoff) {
+        // Update the RC Setpoint/Deflection filter and FeedForward Filter
+        // all cutoffs will be the same, we can optimize :)
+        pt3FilterUpdateCutoff(&smoothingData->filterSetpoint[0], pt3FilterGain(setpointCutoffFrequency, dT));
+        const float pt3K = smoothingData->filterSetpoint[0].k;
```
```suggestion
        const float pt3K = pt3FilterGain(setpointCutoffFrequency, dT);
```

Quick-Flash replied:
Yeah this is the right way to do it, lol.

ledvinap commented on src/main/fc/rc.c:
```diff
@@ -341,71 +341,65 @@ static FAST_CODE_NOINLINE void rcSmoothingSetFilterCutoffs(rcSmoothingFilter_t *
+        // all cutoffs will be the same, we can optimize :)
+        pt3FilterUpdateCutoff(&smoothingData->filterSetpoint[0], pt3FilterGain(setpointCutoffFrequency, dT));
+        const float pt3K = smoothingData->filterSetpoint[0].k;
+        for (int i = FD_ROLL; i <= FD_YAW; i++) {
+            pt3FilterUpdateCutoff(&smoothingData->filterSetpoint[i], pt3K);
+            pt3FilterUpdateCutoff(&smoothingData->filterFeedforward[i], pt3K);
+        }
+        for (int i = FD_ROLL; i < FD_YAW; i++) {
+            pt3FilterUpdateCutoff(&smoothingData->filterRcDeflection[i], pt3K);
         }
```
```suggestion
        for (int i = 0; i <= FD_YAW; i++) {
            pt3FilterUpdateCutoff(&smoothingData->filterSetpoint[i], pt3K);
            pt3FilterUpdateCutoff(&smoothingData->filterFeedforward[i], pt3K);
            pt3FilterUpdateCutoff(&smoothingData->filterRcDeflection[i], pt3K);
        }
```

Quick-Flash replied:
RC deflection only hold pitch and roll. Its an artsy of size 2. This will index outside of the rc deflection array.

ledvinap commented on src/main/fc/rc.c:
```diff
@@ -341,71 +341,65 @@ static FAST_CODE_NOINLINE void rcSmoothingSetFilterCutoffs(rcSmoothingFilter_t *
+
     DEBUG_SET(DEBUG_RC_SMOOTHING, 1, smoothingData->setpointCutoffFrequency);
-    DEBUG_SET(DEBUG_RC_SMOOTHING, 2, smoothingData->feedforwardCutoffFrequency);
+    DEBUG_SET(DEBUG_RC_SMOOTHING, 2, 0);
```
maybe throttleCutoffFrequency ?

ledvinap commented on src/main/fc/rc.c:
```diff
@@ -621,14 +612,15 @@ static FAST_CODE_NOINLINE void calculateFeedforward(const pidRuntime_t *pid, fli
 
     // smooth the setpointSpeed value
-    setpointSpeed = prevSetpointSpeed[axis] + pid->feedforwardSmoothFactor * (setpointSpeed - prevSetpointSpeed[axis]);
+    const float pt1_k = pt1FilterGainFromDelay(pid->feedforwardSmoothFactor, 1.0f / rxRate);
+    setpointSpeed = prevSetpointSpeed[axis] + pt1_k * (setpointSpeed - prevSetpointSpeed[axis]);
```
This is PT1 filter. It would be great to refactor it using filter functions.

ledvinap reviewed (approved):
See comments ..

ledvinap commented on src/main/fc/rc.c:
```diff
@@ -341,71 +341,65 @@ static FAST_CODE_NOINLINE void rcSmoothingSetFilterCutoffs(rcSmoothingFilter_t *
+            pt3FilterUpdateCutoff(&smoothingData->filterSetpoint[i], pt3K);
+            pt3FilterUpdateCutoff(&smoothingData->filterFeedforward[i], pt3K);
+        }
+        for (int i = FD_ROLL; i < FD_YAW; i++) {
```
Maybe:
```suggestion
        for (int i = FD_ROLL; i <= FD_PITCH; i++) {
```

ctzsnooze commented:
Thanks for this PR - very grateful that this old code is being reviewed.

When the RC smoothing and feedforward smoothing code was being developed, we didn't really know whether the same cutoffs could be applied to each parameter.  Now there has been several years where the auto cutoffs are widely used, and the same cutoff has been applied to each of setpoint, throttle and feedforward, without problems.

Of the three parameters, feedforward is the most problematic since it takes the first and second derivatives of setpoint and therefore amplifies any flat spots (eg ELRS telemetry packets) or signal noise.  Hence there is a lot more code required for feedforward smoothing than for setpoint smoothing.

RC smoothing filter cutoffs must adapt to the detected Rx Link Rate, and link rate can change in-flight, so we need to ensure that this still happens.  In the opening comment to the PR, it says "Also fix the issue where the feedforward_smooth_factor would be dependent on the rx rate" - I'm not sure what that means with respect to feedforward.  Stronger feedforward smoothing *is* required at higher link rates, so we do need to change it when the link rate changes.  Feedforward smoothing is, effectively, the RC smoothing of the derivative steps, and is definitely needed.

Is it possible to provide logs which show the raw RC values, un-smoothed setpoint, and smoothed setpoint and feedforward?  So that we compare the results with the revised code, vs the current calculations?

ctzsnooze commented on src/main/flight/pid_init.c:
```diff
@@ -549,7 +549,12 @@ void pidInitConfig(const pidProfile_t *pidProfile)
+    float scaledSmoothFactor = 0.01f * pidProfile->feedforward_smooth_factor;
+    float rxDt = 1.0f / 250.0f;
+    float feedforwardSmoothingTau = (rxDt * scaledSmoothFactor) / (1.0f - scaledSmoothFactor);
+    pidRuntime.feedforwardSmoothFactor = feedforwardSmoothingTau;
```
Why do we re-calculate this `scaledSmoothFactor` value every loop, when it is constant?  
`rxDt` is constant (always 1.250)
`scaledSmoothFactor` is determined by the PID Profile value.
So... if we want to have a 'constant' pidRuntime value for feedforwardSmoothFactor, we should calculate that in pid_init.c, like we do now.

Quick-Flash replied:
This is in the pid_init.c file. This is only calculated once.

ctzsnooze commented:
@Quick-Flash - sorry I took a long time to check the code and had to edit my previous comments.  I still haven't finished but I have checked the feedforward aspects.

Of course, RC_smoothing just 'interpolates smoothly between the steps' in both setpoint, feedforward and throttle.

For setpoint, there is no 'manipulation' of the steps, the RC_smoothing just draws something like a straight line between those steps.

You said, "(it's actually funny as with the current code logic the faster looprate you run the higher value you need to set the rcsmoothing, which feels odd to me)".  

Let's not use the term 'looprate' here.  It's the **RX Link Rate** which determines how much, or how little, RC_smoothing is needed to optimally smooth out the sharp steps.

And if those steps arrive more frequently, then the 'slope' of the 'interpolation line' that is drawn by the PT3 filter should be 'faster', since the steps are all smaller, and coming in more frequently.

Hence the *cutoff value* of the PT3 filter is a higher value at higher Rx Link Rates, but the 'smoothing' effect of a higher frequency filter cutoff is *less*.

So yes filters are always counterintuitive; higher values result in less smoothing, and vice versa, which is counterintuitive.

(In part that's why I chose  a smoothing factor, as in `feedforward_smooth_factor`, where a value of 0 resulted in no smoothing, and the higher the value, the more smoothing you got, for smoothing the steps in feedforward.)

The delay from automatic RC_smoothing always decreases with faster link rates, because the code automatically chooses a higher RC_smoothing frequency (and lower latency) for higher link rates.  The user doesn't have to do anything, and they get the optimal balance between delay and smoothing.

Unfortunately, we didn't automate the smoothing for feedforward.

This PR does that, and that's a great step forward in user convenience.  It just needs the default `feedforward_smooth_factor to be set to 50, and a lot of old Presets to be updated.  @sugaarK @limonspb pls take not of that.

For feedforward, we first apply smoothing and averaging *to the steps* at Rx Link Rate, and then once we have 'smoothed' the FF steps themselves, we submit that data to RC_smoothing to have the sharp edges smoothed out.  This PR will simplify the user experience when setting the step smoothing factor for feedforward; and, do so dynamically.

The only technical issues I see with this otherwise very useful PR are:

- pidRuntime.feedforwardSmoothFactor should be calculated in pid_init.c, once, not every time we get a new packet in rc.c
- the default value for `feedforward_smooth_factor` needs to be changed to 50
- all Rx Link Presets should have their `feedforward_smooth_factor` values changed to 50

.

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -340,72 +340,48 @@ bool getRxRateValid(void)
-        }
+
+    // Update the RC Setpoint/Deflection filter and FeedForward Filter
+    // all cutoffs will be the same, we can optimize :)
```
Here we are updating filters of two types for three axes every time we get a new packet.  We also do an update for the feedforward smoothing cutoff as well.  That's quite a few updates, within for loops, even though they are simplified a bit.

In `processRcSmoothingFilter(void)` under `isRxDataNew` we decide whether there has been a significant change in interval.  The code here was optimised a good while back to avoid unnecessary filter changes when the Rx link had occasional dropouts, to handle gradual small changes, and to respond reasonably quickly to a true change in Rx Link frequency.  If the update criteria were met, the cutoffs were re-calculated.

Elsewhere where we updated the filters, we checked if those cutoffs had changed (eg  line 358 of the old code), and if so, we updated the filters.

That code may not be ideal but it should be used and work the same in the context of this PR.  Probably there is a nicer way to code this than what we did, but the functionality we had should not change here; if we want to the code that decides whether these cutoffs should change that should be done in a different PR, and tested carefully with a large number of receivers and with packet loss situations

Quick-Flash replied:
> In addition, the change comparison wanted more than 1Hz of difference between every third sample of the smoothed link speed before it would change the cutoffs.

The code previously and now doesn't care how much the rcSmoothingData.smoothedRxRateHz has changed before checking to see if the cutoffs should change, but yes it is only after every 3 new samples that it will run this check. That rcSmoothingData.smoothedRxRateHz also is only updated every 3 new valid packets, FF uses this along with other stuff.

> I accept that the 1Hz delta test is a dodgy way to look for a significant change, especially when the link speed is close to the boundary. But it does ensure that any change greater than 1Hz from the last used frequency will be actioned.

Updating it every time is also a way to ensure that any change greater than 1hz will be actioned.

> Maybe confirm that, in the old method, the test:
(smoothingData->setpointCutoffFrequency != oldSetpointCutoff)
was only true when the setpointCutoffFrequency changed, and this was only after every three valid link samples.
That's because rcSmoothingSetFilterCutoffs() only ran after every third valid packet was received, excluding outliers. Hence smoothingData->setpointCutoffFrequency only changed every third valid sample.

Well yes, that it how it worked. Now it works the same except without a check to see if the old setpointcutoff and new setpointcutoff are the same value (or close to the same)
@ctzsnooze

ctzsnooze replied:
@Quick-Flash 

To summarise the previous behaviour:
- the rcSmoothing code looked at each RC packet interval, excluded aberrant values, smoothed them with a PT1, and after every third value, sent a request to update the filters via `rcSmoothingSetFilterCutoffs()`.
- `rcSmoothingSetFilterCutoffs` then ignored 'small' changes which didn't exceed one integer difference from the previous cutoff.  
- in the feedforward code, there was no adaptation for packet interval.  Instead a fixed user-defined filter 'k' value was applied for all steps.  However, the 'setpoint speed' value was calculated based on the reported individual Rx Rate value for each step.  So in essence the 'filter' *was* updated every iteration. 

Moving forward with this PR:
- this PR should retain the same Rx Rate smoothing and 'every third value' code for RC smoothing, and does.
- the value of ignoring changes within the same integer value is uncertain.  On 500Hz links, auto cutoffs are around 200Hz, even if we acted on every change, a change of up to 1Hz out of 200hz is a very small change in cutoff and would cause only the tiniest glitch in output; if the change was more than 1Hz it would get through.  At 250Hz it's still only less than a 1% change in cutoff, and hence less than 1% of a jump at the output.  At 50Hz the percentage link interval variation is much less.  So yes I tend to agree with removing that integer check.
- in the feedforward code, we previously did adapt the 'setpoint speed' for the actual packet intervals.  So therefore, to change nothing with a filter model, we *should* update the feedforward step smoothing filter cutoff with the current Rx Rate.  I think the PR achieves this.

ctzsnooze commented:
From memory the old default `feedforward_smooth_factor` of 25 was to handle, sort of, old FrSky at 90Hz / TBS at 150Hz and not be too awful at 250Hz.  It was an old, old default, and was subsequently handled with Presets for each link speed, so it didn't matter much.

What you've done here is way better, and works great with the default of 50.

However, there is a possibly unintentional change here in that each filter cutoff and k value is updated on the instantaneous interval value.  This caused issues in the past where samples were dropped, or the intervals we detected had 'false' variability in them.  Then the cutoff for the filter suddenly changed, changing the output value.  If the next interval compensated the other way, there would be a up/down amplitude variation in the output.  Hence in `processRcSmoothingFilter()` we have some complex and carefully tested code that excludes 'extreme` values, gets a smoothed average of recent values, and then only updates the cutoffs every third valid sample from the smoothed average.

Probably this can be re-factored in a better way, but I do recall that this code was tested with a large number of Tx devices and found to effectively handle a range of issues relating to packet intervals.

Hence let's please not change how that is done in this PR.  On the other hand, if that code can be effectively refactored, eg perhaps with some variable `updateFiltersNow` or similar, that would be excellent.

ctzsnooze commented:
Also, the filters must be initialised at some point.  With auto smoothing, we don't set `rcSmoothingData.filterInitialized = true` until we have three valid stable values.  Until then I don't think the filter can run properly, but I don't really know for sure.
In this PR the checks for the filter being initialised before updating the cutoffs have been removed.
Will that cause problems?
@KarateBrot I'd appreciate your thoughts here.

Quick-Flash commented on src/main/fc/rc.c:
```diff
@@ -621,15 +594,18 @@ static FAST_CODE_NOINLINE void calculateFeedforward(const pidRuntime_t *pid, fli
     // smooth the setpointSpeed value
-    setpointSpeed = prevSetpointSpeed[axis] + pid->feedforwardSmoothFactor * (setpointSpeed - prevSetpointSpeed[axis]);
+    const float pt1K = pt1FilterGainFromDelay(pid->feedforwardSmoothFactor, 1.0f / rxRate);
+    pt1FilterUpdateCutoff(&rcSmoothingData.filterSetpointSpeed[axis], pt1K);
```
Most the filters still use that logic, just not the 1 filter. Also, the algorithm for cutoff change is filtered and a single missed packet will do little to effect it. But this logic could be moved there, it'll then only be updated once instead of 3 times (that k is the same for all axis).

Quick-Flash commented:
> Also, the filters must be initialised at some point.  With auto smoothing, we don't set `rcSmoothingData.filterInitialized = true` until we have three valid stable values.  Until then I don't think the filter can run properly, but I don't really know for sure.
> In this PR the checks for the filter being initialised before updating the cutoffs have been removed.
> Will that cause problems?
> @KarateBrot I'd appreciate your thoughts here.

Not checking for 3 values before calling it stable will never cause an issue. No one will ever arm their drone before they have received 3 valid packets.

ctzsnooze commented on src/main/flight/pid_init.c:
```diff
@@ -549,7 +549,12 @@ void pidInitConfig(const pidProfile_t *pidProfile)
     pidRuntime.feedforwardTransitionInv = (pidProfile->feedforward_transition == 0) ? 0.0f : 100.0f / pidProfile->feedforward_transition;
     pidRuntime.feedforwardAveraging = pidProfile->feedforward_averaging;
-    pidRuntime.feedforwardSmoothFactor = 1.0f - (0.01f * pidProfile->feedforward_smooth_factor);
+    // feedforward_smooth_factor effect previously would change based on packet looprate
```
No... although I agree with this part:
`feedforwardSmoothingTau = (0.004 * 0.65) / 0.35 = 0.0026 / 0.35 = 0.0075`
I don't agree with the second paet.
The dT used in pt1FilterGainFromDelay is from the present Rx interval, ie RxRate, which is `currentRxRateHz = 1e6f / currentRxIntervalUs;` so at 500Hz
```
return dT / (dT + delay)
= 0.002 / (0.002 + 0.0075)
= 0.002 / 0.0095
= 0.21
```
It has to be the current Rx interval as rate, otherwise there is no normalisation.

Quick-Flash commented:
> My concern wasn't stability or when a person armed it. It was that the filter init code must run before any attempt is made to use the filter, otherwise it isn't set up right. I'm sure we can start setting variables to cutoff values, but we can't do a filterApply if the filter code hasn't yet been initiated. That was my concern.

Ah, yeah I see looking over this again, init is never called for this filter. Which the problem is really just that the state variables are not all set to 0, so they could potentially be any value to start which might possibly cause an issue (unless it is NAN of +-INF it would be OK (There will be enough time to filter it down to reasonable values, but these numbers would break it)).
We can just init the code with any random cutoff (it'll be overwritten quickly anyhow) and just ensure that all the state is initialized to 0. I'll do that and change the default ffsmoothing thing to a value of 50 in a commit.

ctzsnooze commented:
> We can just init the code with any random cutoff...

A value of 50Hz for the 'random' cutoff might be sufficient to smooth some erratic behaviour at the start?

Quick-Flash commented:
> > We can just init the code with any random cutoff...
> 
> A value of 50Hz for the 'random' cutoff might be sufficient to smooth some erratic behaviour at the start?

The value does not matter at all. It will be overwritten very very quickly. What actually matters is initializing the state to 0 nothing else as before it's ever used we update the cutoff. 

 
> Additionally, we absolutely must do some A-B comparisons of RC smoothing and feedforward behaviour before and after this PR, with a variety of Rx systems, before merging.

Some testing sure, a variety of Rx systems just isn't needed when this just largely refactors existing logic to act more or less the same. So there just honestly isn't a need for lots of testing of different gear.

Quick-Flash commented:
> A value of 50Hz for the 'random' cutoff might be sufficient to smooth some erratic behaviour at the start?
> 
> @Quick-Flash - however, this may cause a problem with logging.
> 
> The logging header currently shows the auto-determined filter cutoff values.  There were some issues making this work properly that were resolved only relatively recently.
> 
> The previous code would not set any auto-determined filter value until it had received 'valid' Rx intervals, and then after averaging three such values.  That was because the filter was not initialised *at all* until we had some reasonable estimate of the real cutoff that we wanted to use.
> 
> If we initialise at some 'random' cutoff, regardless of the incoming data, we would need to validate that the correct values for cutoff still appear in the header of a log.

I'm guessing that when logging starts it grabs the current cutoff. So I don't think this behaviour will have changed, but we'd need to double check.

Quick-Flash commented:
> @Quick-Flash - is this PR ready for detailed bench testing at this point? Or should I wait for a commit that handles filter initialisation? In regard to initialisation, an interval consistent with 100Hz would be a decent value, and would make it easy to confirm that the value in the log header is correct (by testing 50Hz and 500Hz links).
> 
> Previously the testing I did was to deliberately change frequency and log RC packets and the reported intervals, and also check feedforward was appropriate.
> 
> It's a big task to do this, so I don't want to do it until you say, OK, this is ready for testing.

You can check my latest commit, but I have gotten this to initialize on startup (it just assumes a 100hz rc rate at start).
This should be ready to go now. Best of luck!

ctzsnooze commented:
@Quick-Flash - could you possibly include:
```
    DEBUG_SET(DEBUG_RX_TIMING, 5, smoothingData->setpointCutoffFrequency);
```
with the other logging of the setpoint cutoff around line 375 of rc.c?

That slot is currently unused, and lets us see the change in setpoint cutoff as the link changes frequency or has dropouts.  I'll do this locally for testing, but we should do it in this PR also I think.

ctzsnooze commented:
Otherwise... no problems at all with the PR.  Functionally, it seems identical to Master, in testing with auto smoothing enabled.

However... there are some issues we could perhaps tidy up :-) mostly at 500 and 1000hz.

In these tests I:
- compared master to this PR
- disabled all feedforward step averaging and smoothing, so the yellow feedforward trace in the top panel shows only RC smoothing.
- added logging to debug 5 (orange) of the RC Setpoint Smoothing cutoff, to see how it changed when the link changed.
- on the bench, pushed roll back and forwards while changing frequencies from 50 to 1000Hz on ELRS with 1:32 telemetry on (meaning no data received every 32 packets)

Notable findings:
- no problems associated with this PR
- the setpoint smoothing filter changes much more than it should at higher link speeds, probably because the variability in time interval between packets is relatively larger in proportion to the average value of the interval.
- hence at 1000Hz there is constant changing of the setpoint filter cutoff, with about 10% variation; there is about 15% variation in interval; at 500Hz there are random, occasional drops in cutoff, but only about 4-5% variation in cutoff and interval; at 250Hz, the frame delta is less than 1% and the cutoff almost never changes - see images below.  The code should do better than this - we should not be constantly changing cutoff at higher link speeds.
- at 1000Hz the `isRateValid` flag (yellow debug trace in lower panel) is repeatedly triggered.  This probably associates with the proportionally more significant link variability.  However these flags are false; the link is still active, and working normally.  This should definitely be fixed.  It doesn't cause false RXLoss indication, but only because `RXLOSS` requires a longish period of time without any valid intervals.  Probably arises because at 1000Hz, there are a lot of packets with 830us intervals, less than the 950us upper limit, and I think that being outside Min or Max will trigger the `isRateValid` flag.
- behaviour while changing rate is acceptable, and similar in master and the PR; no erratic or crazy values in the filters or the link estimates.
- `currentRxIntervalUs` is not different from `delta`
-  constraining of frame delta to not less than 950us cleans up the 1000Hz trace a little bit.

Note that the debug fields in blackbox log explorer are not named correctly, so in these images I show all 8 fields, summarised as:
0 - red  - **frame `delta`** in ms (raw interval value), constrained
1 - cyan - frame age timer (ignore this one, it just ramps up)
2 - yellow - `isRateValid` flag (should be 1; drops a lot at 1000F, otherwise stable)
3 - purple - `currentRxIntervalUs` - should be 'smoothed' with outliers removed
4 - blue - `delta` unconstrained
5 - orange - **setpoint cutoff**, ie the RC filter cutoff.
6 - green - link quality (should be 100, always is, except when changing link speed)
7 - pink - `isRxReceivingSignal` flag (never triggers in these tests), triggers RXLOSS

First image is 1000Hz, PR1441 on left, master on right.
No notable differences.
In both cases:
- lots of noise in FF (absolutely must have appropriate FF step averaging and smoothing set!)
- about 15% variability in delta and cutoff
- cutoff changes typically around 4Hz but at times at 2Hz.
- filter frequency set to around 360Hz in both cases.
<img width="1750" height="1025" alt="Screenshot 2025-08-09 at 4 12 09 pm" src="https://github.com/user-attachments/assets/b6316444-f5be-4a84-929b-1a581269f104" />

ctzsnooze commented:
Here at 500Hz:
- no difference between PR and Master, as for 1000hz.
- less absolute FF variation, obviously
- setpoint cutoff around 187
- much less percentage variation in setpoint cutoff; intermittent drops (for no obvious reason), by about 5%
- `isRxValid` flag stable at 1
- only about 5% variability in frame delta, usually up/downs.
- setpoint doesn't change much overall

<img width="1748" height="1031" alt="Screenshot 2025-08-09 at 5 03 46 pm" src="https://github.com/user-attachments/assets/eb4bca1b-fb67-4d78-a729-32879cf44ef8" />

ctzsnooze commented:
@Quick-Flash

Here are the logs used for the above tests.

Note that `feedforwardSmoothFactor` is set to 0 in these logs, so I'm not testing that yet..  

In all other respects, the PR appears to not change anything adversely.
[Master_RxLink_NoFFSm.bbl.zip](https://github.com/user-attachments/files/21696733/Master_RxLink_NoFFSm.bbl.zip)

[14411_Rx_Link_NoFFSm_2.bbl.zip](https://github.com/user-attachments/files/21696732/14411_Rx_Link_NoFFSm_2.bbl.zip)

Quick-Flash commented:
>     * the setpoint smoothing filter changes much more than it should at higher link speeds, probably because the variability in time interval between packets is relatively larger in proportion to the average value of the interval.
> 
>     * hence at 1000Hz there is constant changing of the setpoint filter cutoff, with about 10% variation; there is about 15% variation in interval; at 500Hz there are random, occasional drops in cutoff, but only about 4-5% variation in cutoff and interval; at 250Hz, the frame delta is less than 1% and the cutoff almost never changes - see images below.  The code should do better than this - we should not be constantly changing cutoff at higher link speeds.
> 
>     * at 1000Hz the `isRateValid` flag (yellow debug trace in lower panel) is repeatedly triggered.  This probably associates with the proportionally more significant link variability.  However these flags are false; the link is still active, and working normally.  This should definitely be fixed.  It doesn't cause false RXLoss indication, but only because `RXLOSS` requires a longish period of time without any valid intervals.  Probably arises because at 1000Hz, there are a lot of packets with 830us intervals, less than the 950us upper limit, and I think that being outside Min or Max will trigger the `isRateValid` flag.

One issue is that this code will "learn" a new rx rate faster if you are using a faster packet rate. This filter should probably be adapted so that it uses the same cutoff at each frequency. So using the smoothed rx rate hz to update the cutoff of the pt1 filter. This would probably help smooth out some of the jitteryness of the 1000hz rx signals.

```
                        static float prevSmoothedRxRateHz;
                        // smooth the current Rx link frequency estimates
                        const float kF = 0.1f; // first order kind of lowpass smoothing filter coefficient
                        // add one tenth of the new estimate to the smoothed estimate.
                        const float smoothedRxRateHz = prevSmoothedRxRateHz + kF * (currentRxRateHz - prevSmoothedRxRateHz);
                        prevSmoothedRxRateHz = smoothedRxRateHz;
```

As for the isRatevalid you are correct about when it will trigger. We could decrease that min interval to be 800us or so, that may help with this as well.

ctzsnooze commented:
@Quick-Flash - Checking feedforward smoothing now.

With this PR, there is consistent delay for a given smoothing value.

Although this is very different from master, it's probably a good thing.

For example, a smoothing value of 80 returns approximately 16ms of delay (eyeball measurement), regardless of link speed.  pt1K with a feedforward smoothing of 80 changes almost 10-fold over the range of link speeds, from 0.06 at 1000Hz, to 0.20 at 250Hz, and to 0.56 at 50Hz.  

In contrast, master with a feedforward smoothing of 80 has an effective pt1K of 200, regardless of link speed.  Master smooths much more at 50Hz, with four times more delay, is the same at 250Hz, and doesn't smooth much at all at 1000Hz, leading to an excessively noisy trace at 1000Hz.  

At a value of 50, with this PR there probably isn't enough smoothing at 50Hz, and it is definitely too noisy at 1000Hz.

Adding 2 point averaging improves matters but not quite enough.  1000Hz is still too noisy, 250Hz is quite good but not could be smoother, probably, and 50Hz still isn't as smooth as master.

The 'sweet spot' seems to be a smoothing value of 70-75 with 2-point averaging enabled.  That works for most setups.

ctzsnooze commented:
@Quick-Flash 

I agree, lets change the Rx Min to 800us, that should prevent most false triggers for 100Hz users.

Please set the feedforward defaults to feedforward averaging on, and feedforward smoothing of 65. I think most pilots, regardless of link speed, will be happy with this.

ctzsnooze commented:
Images from this morning's testing (almost 6 hours to get there).

Modified a debug to log the 'raw' internal 'setpoint speed' values before (lower panel, blue), and after filtering (lower panel orange), but not including averaging.  Built master and this PR for comparison with the debug changes.

Top panel shows Setpoint, RC command and final feedforward (including boost of 15 and averaging).

The yellow final feedforward trace in the upper panel is 'all that matters'.  Lag can be compared by any delay when the stick crosses the center.  Each panel is 1s.

Here averaging is set to 2 point in all traces.  

The windows are feedforward smoothing of 50, 65 and 75.  The value makes quite a big difference.

First, at 1000Hz.  Probably 65 is a good balance of noise vs responsiveness.

<img width="1795" height="1044" alt="Screenshot 2025-08-10 at 1 15 14 pm" src="https://github.com/user-attachments/assets/21da5feb-0975-40f1-afa7-57722c32012c" />

ctzsnooze commented:
Similar graphs for 500Hz, 333Hz, 250Hz:

First, 500Hz:
<img width="1787" height="1032" alt="Screenshot 2025-08-10 at 1 23 18 pm" src="https://github.com/user-attachments/assets/ba821f98-681c-4177-bcb7-91c36d91a5ba" />

then 333Hz :
<img width="1776" height="1026" alt="Screenshot 2025-08-10 at 1 21 55 pm" src="https://github.com/user-attachments/assets/b7dd2c7f-5d69-4570-aba1-cea477017fb7" />

and 250Hz:
<img width="1798" height="1035" alt="Screenshot 2025-08-10 at 1 25 21 pm" src="https://github.com/user-attachments/assets/f361efaa-20f0-48bd-a727-05464f16c4d2" />

ctzsnooze commented:
@Quick-Flash - one last comment: I honestly think we are re-calculating the cutoffs far too frequently. 

The green trace on the above logs, which tracks `pt1K`, the filter cutoff, shows that we are re-calculating values that are very much the same, all the time, on all three axes.  

Also, there are 'spikes' in the cutoff, whenever ELRS sends a telemetry packet.  The `pt1K` value transiently goes higher by about 50% for that interval, meaning that feedforward more rapidly responds to the new data that arrives.

This is probably a good thing, so long as the filter drops back to normal immediately after.

Feedforward itself adapts to these situations quite well, ie there is no 'spike' in the underlying feedforward value.

I don't really mind doing these recalculations so often, but I'm not convinced that it is helpful.  The small variations in interval are proportionally great at higher link speeds.  I think they arise not from the Tx, which has an accurate clock, but from our loop time variability.  

Hence I expect that re-calculating the filters for small changes isn't likely to be helpful.  

Image below shows no averaging, feedforward smoothing of 80, Master on left, PR on right, 500Hz link.  Green trace shows Pt1K.

Master has the fixed pt1K of 0.2, PR14411 has 0.115 mostly and 200 during telemetry packets.

The variability in pt1K is about 8% (more at 1000Hz). Do you agree that this adds to FF noise?  Note that the end result is not directly comparable, noise wise, in these logs - it's more of a 'theoretical' question.

<img width="1740" height="1068" alt="Screenshot 2025-08-10 at 2 03 30 pm" src="https://github.com/user-attachments/assets/5b5d4422-8e5b-4e60-b2e2-d3019e85e0c7" />

ctzsnooze commented:
I wrote some code that only did the feedforward pt1K updates when there was more than a 20% change for 3 consecutive rxIntervals.  This is the same coding approach used in rcSmoothing.

Here is a comparison of the results at 500hz
- cutoff is set automatically, and fixed until the link speed is changed
- no discernible difference in discernible noise in FF at any speed
- no need to recompute and update the cutoff unless link speed changes

Still to early to be sure - constant cutoff version on right.

500Hz:
<img width="1782" height="1030" alt="Screenshot 2025-08-11 at 12 00 14 am" src="https://github.com/user-attachments/assets/d55e3077-e8eb-4b6c-905d-32ab4186533d" />

1000Hz:
<img width="1774" height="1057" alt="Screenshot 2025-08-11 at 12 04 18 am" src="https://github.com/user-attachments/assets/d8ea8da6-c6f1-462b-aae2-acec1230e1f4" />

Definitely looks like using an interval that is not very dynamic gives acceptable results.

Here is a transition from 250 to 333Hz, using the above algorithm, showing 3s of data:

<img width="1728" height="1036" alt="Screenshot 2025-08-11 at 12 09 12 am" src="https://github.com/user-attachments/assets/f8ae04f3-aa39-4f18-807a-2609ab65c6cc" />

Quick-Flash commented:
> `ifdef RCSMOOTHING`

I don't believe so, but I am also wondering what the usefulness of having the `#ifdef USE_RC_SMOOTHING_FILTER` in the first place is. I mean is it something we would ever want to disable for any real reason?

Quick-Flash commented:
> 65

Can do, you asked for a default of 50 earlier, but I can see that with your testing 65 seems like a better default.

Quick-Flash commented:
> The only thought I have is the 15Hz hard-min cutoff - for users who are trying to do the 'cinematic FF forcing with extremely low cutoff' as a crutch for flying ability (me) - is there any reason we couldn't go arbitrarily low on that (or at least down to 5Hz - which is garbage to the point of negating FF unless we're talking large craft, or maybe wings)?

I don't see why not, I don't even know you can in general tweak the smoothing enough to hit those values at a very high rx rate. But I don't hate the idea of doing that.

Quick-Flash commented:
@ctzsnooze I've just gone and taken the code you gave me and modified it slightly. I think that this should be pretty good now. I also made the FF jitter no longer care what looprate you are running at. Also the rcsmoothing samplecount is now gone, so we can use those debug slots for something else.

Also please look this over, I did try to simplify it a bit and split a few things out into separate functions to make it easier to read.

Quick-Flash commented on src/main/fc/rc.c:
```diff
@@ -415,84 +396,32 @@ static FAST_CODE void processRcSmoothingFilter(void)
 
         rcSmoothingData.autoSmoothnessFactorSetpoint = 1.5f / (1.0f + (rxConfig()->rc_smoothing_auto_factor_rpy / 10.0f));
-        rcSmoothingData.autoSmoothnessFactorFeedforward = 1.5f / (1.0f + (rxConfig()->rc_smoothing_auto_factor_rpy / 10.0f));
         rcSmoothingData.autoSmoothnessFactorThrottle = 1.5f / (1.0f + (rxConfig()->rc_smoothing_auto_factor_throttle / 10.0f));
 
         rcSmoothingData.setpointCutoffSetting = rxConfig()->rc_smoothing_setpoint_cutoff;
         rcSmoothingData.throttleCutoffSetting = rxConfig()->rc_smoothing_throttle_cutoff;
-        rcSmoothingData.feedforwardCutoffSetting = rxConfig()->rc_smoothing_feedforward_cutoff;
 
         rcSmoothingData.setpointCutoffFrequency = rcSmoothingData.setpointCutoffSetting;
-        rcSmoothingData.feedforwardCutoffFrequency = rcSmoothingData.feedforwardCutoffSetting;
         rcSmoothingData.throttleCutoffFrequency = rcSmoothingData.throttleCutoffSetting;
 
         if (rxConfig()->rc_smoothing_mode) {
-            calculateCutoffs = rcSmoothingAutoCalculate();
-            // if we don't need to calculate cutoffs dynamically then the filters can be initialized now
-            if (!calculateCutoffs) {
-                rcSmoothingSetFilterCutoffs(&rcSmoothingData);
-                rcSmoothingData.filterInitialized = true;
-            }
+            // shouldRecalculateCutoffs is handle here so that rc_smoothing_mode overwrites other settings
+            rcSmoothingData.shouldRecalculateCutoffs = rcSmoothingAutoCalculate();
+            rcSmoothingSetFilterCutoffs(&rcSmoothingData);
+            rcSmoothingData.filterInitialized = true;
         }
```
Good idea

Quick-Flash commented on src/main/fc/rc.h:
```diff
@@ -29,6 +29,18 @@
+    float prevRxInterval[XYZ_AXIS_COUNT];
+    pt1Filter_t filterSetpointSpeed[XYZ_AXIS_COUNT];
+    pt1Filter_t filterSetpointDelta[XYZ_AXIS_COUNT];
+} feedforwardData_t;
```
Is this really a big deal? won't we just handle it once it is a problem?

ctzsnooze commented on src/main/fc/rc.c:
```diff
@@ -704,10 +604,80 @@ static FAST_CODE_NOINLINE void calculateFeedforward(const pidRuntime_t *pid, fli
+
+    const bool ready = (targetPidLooptime > 0);
+
+    if (ready && isRxReceivingSignal() && isRxRateValid) {
```
Strongly recommend removing the `ready` check altogether.  It's just a waste of CPU cycles.
I've tested removing it, and there is no difference in the logs.
- the correct link frequency is still stored in the log header.
- the initial frequency on arming is perfectly good.
- the initial frequency on powering up and arming as soon as possible is also perfectly good.
- even if the initial frequency was incorrect, it cannot be extreme (must still be 'valid')
- even if the initial frequency is very wrong, the code quickly homes in on the correct frequency.

ctzsnooze replied:
Great,  nice to see it removed!

Quick-Flash commented on src/main/fc/rc.c:
```diff
@@ -704,10 +599,81 @@ static FAST_CODE_NOINLINE void calculateFeedforward(const pidRuntime_t *pid, fli
+
 FAST_CODE void processRcCommand(void)
 {
+	bool updateSmoothing = false;
     if (isRxDataNew) {
+        updateSmoothing = shouldUpdateSmoothing();
+
+        if (updateSmoothing) {
+#ifdef USE_FEEDFORWARD
+            // update FF smoothing
+            updateFeedforwardFilters(&pidRuntime);
+#endif
+#ifdef USE_RC_SMOOTHING_FILTER
+            if (rcSmoothingData.shouldRecalculateCutoffs) {
+                // for auto calculated filters, calculate the link interval and update the RC smoothing filters at regular intervals
+                // this is more efficient than monitoring for significant changes and making comparisons to decide whether to update the filter
+                rcSmoothingSetFilterCutoffs(&rcSmoothingData);
+                DEBUG_SET(DEBUG_RC_SMOOTHING_RATE, 2, smoothedRxRateHz); // only changes curing these conditions
+            }
+#endif
+        }
```
> Yes that's a good observation. If the profile is changed, we need to update the smoothing settings for feedforward.

Which happens as soon as we get enough valid rx signals to change settings.

ctzsnooze commented:
@KarateBrot @ledvinap @Quick-Flash 

I've tested this PR very extensively, at least 10 hours of testing.

It works really well.

Adaptation to changes in Rx Link interval is clean, and the logic is much clearer:
- identify 'outlier' packets as being those more than 12.5% different in interval from the smoothed average value
- if there are more than three 'outliers' in a row, assume that the true link rate has changed, and shift to that frequency
- for packets that are not outliers, keep a PT1 type smoothed link rate value (k of 0.1 at link speed) to get a reasonably smooth average of the 'valid' intervals (ie, excluding telemetry intervals that are twice as long as usual, etc)
- record the smoothed link rate value in the blackbox header
- re-calculate both the feedforward smoothing and RC smoothing cutoffs every three non-outlier packets, to the smoothed link rate value

The same logic is therefore used for both feedforward cutoff updating and rc smoothing cutoff updating.  It's now in a dedicated function.

 My only suggestion is that I don't think we need the 'ready` check.  I've tested without it and there are no problems without it.  Not that we would expect any.
 
 I'll review the debugs needed now the code is, I think, ready to go.
 
 This is a significant improvement on the previous situation, since feedforward smoothing now works consistently across the full range of frequencies (50 - 1000Hz) with the same default setting of 2-point averaging and a smoothing factor of 65.
 
 The only users who may want to increase the feedforward smoothing factor are:
 - 1000Hz users who want to lower the jitter reduction value, or just have a smoother trace; a value of 75-80 will be noticeably smoother at the cost of a little more delay.
 - 50Hz users who fly cinematic will find that feedforward is now more apparent than before, since there is effectively less smoothing on defaults.  They may prefer a feedforward smoothing value of 75-80.
 
 Otherwise the defaults proposed here are very usable at essentially all link speeds.
 
 I'd be grateful if anyone testing the PR could try 1000Hz and 50Hz vs their normal link speed.

Quick-Flash commented:
@coderabbit can you implement your two nitpick comments?

Quick-Flash commented on src/main/fc/rc.c:
```diff
@@ -947,7 +891,27 @@ void initRcProcessing(void)
+    }
+#endif
+
 #ifdef USE_FEEDFORWARD
+    updateFeedforwardFilters(&pidRuntime);
     feedforwardAveraging = pidRuntime.feedforwardAveraging;
     pt1FilterInit(&feedforwardYawHoldLpf, 0.0f);
 #endif // USE_FEEDFORWARD
```
Then please approve this and stop complaining about this line. As I said before earlier code handles the buffer capacity concerns, lets just ignore this and move on.

ledvinap commented:
@sugaarK : I don't see obvious problems. This PR did simplify it a bit and removed some LOC  (+257 −312).

It is very hard to review PRs in these parts of code - there is so much that can be improved/fixed, but unrelated to actual PR.
